packages/features/ee/billing/credit-service.ts

```diff
@@ -30,100 +65,132 @@ export class CreditService {
     bookingUid?: string;
     smsSid?: string;
   }) {
-    let teamIdToCharge = credits === 0 && teamId ? teamId : undefined;
-    let creditType: CreditType = CreditType.ADDITIONAL;
-    let remainingCredits;
-    let userIdToCharge;
-    if (!teamIdToCharge) {
-      const result = await this.getUserOrTeamToCharge({
-        credits: credits ?? 1, // if we don't have exact credits, we check for at east 1 credit available
-        userId,
-        teamId,
-      });
-      teamIdToCharge = result?.teamId;
-      userIdToCharge = result?.userId;
-      creditType = result?.creditType ?? creditType;
-      remainingCredits = result?.remainingCredits;
-    }
+    return await prisma
+      .$transaction(async (tx) => {
+        let teamIdToCharge = credits === 0 && teamId ? teamId : undefined;
+        let creditType: CreditType = CreditType.ADDITIONAL;
+        let remainingCredits;
+        let userIdToCharge;
+        if (!teamIdToCharge) {
+          const result = await this._getUserOrTeamToCharge({
+            credits: credits ?? 1, // if we don't have exact credits, we check for at east 1 credit available
+            userId,
+            teamId,
+            tx,
+          });
+          teamIdToCharge = result?.teamId;
+          userIdToCharge = result?.userId;
+          creditType = result?.creditType ?? creditType;
+          remainingCredits = result?.remainingCredits;
+        }
 
-    if (!teamIdToCharge && !userIdToCharge) {
-      log.error("No team or user found to charge. No credit expense log created");
-      return null;
-    }
+        if (!teamIdToCharge && !userIdToCharge) {
+          log.error("No team or user found to charge. No credit expense log created");
+          return null;
+        }
 
-    await this.createExpenseLog({
-      bookingUid,
-      smsSid,
-      teamId: teamIdToCharge,
-      userId: userIdToCharge,
-      credits,
-      creditType,
-    });
+        await this._createExpenseLog({
+          bookingUid,
+          smsSid,
+          teamId: teamIdToCharge,
+          userId: userIdToCharge,
+          credits,
+          creditType,
+          tx,
+        });
 
-    if (credits) {
-      await this.handleLowCreditBalance({
-        teamId: teamIdToCharge,
-        userId: userIdToCharge,
-        remainingCredits: remainingCredits ?? 0,
-      });
-    }
+        let lowCreditBalanceResult = null;
+        if (credits) {
+          lowCreditBalanceResult = await this._handleLowCreditBalance({
+            teamId: teamIdToCharge,
+            userId: userIdToCharge,
+            remainingCredits: remainingCredits ?? 0,
+            tx,
+          });
+        }
 
-    return { teamId: teamIdToCharge, userId: userIdToCharge };
+        return {
+          teamId: teamIdToCharge,
+          userId: userIdToCharge,
+          lowCreditBalanceResult,
+        };
+      })
+      .then(async (result) => {
```

Comment:
I wanted to have the email sending logic out of the db transaction.

So I've extracted the email sending logic from `_handleLowCreditBalance`, and we pass the result to `handleLowCreditBalanceResult` function. Now, `_handleLowCreditBalance` function just performs database related logics within a transaction, and we run the email sending logic AFTER the transaction is (successfully) done.